# Домашнее задание №6.2
<hr>

#### Текст домашнего задания

Реализовать с использованием потоков и процессов скачивание файлов из интернета.  
Список файлов для скачивания подготовить самостоятельно (например изображений, не менее 100 изображений или других объектов).  
Сравнить производительность с последовательным методом.  
Сравнивть производительность Thread и multiprocessing решений.  
Попробовать подобрать оптимальное число потоков/процессов. 

#### Анализ задания


1. Найти и подготовить список файлов. (https://loremflickr.com/160/160 возьмём картинки здесь)
2. Реализовать скачивание картинок из интернета с помощью threading, multiprocessing и последовательное скачивание.
3. Расчитать время выполнения функций скачивания.
4. Сравнить производительность разных методов скачивания.

#### 1. Импорт необходимых для работы модулей

In [1]:
import requests # библиотека, которая облегчает выполнение HTTP-запросов
import time
import concurrent.futures
import threading

#### 2. Создаём список ссылок и функцию для скачивания картинки.

In [2]:
images = ['https://loremflickr.com/160/160']*100

def download_image(url, index): # cкачиваем изображение по URL и сохраняем его в папку images
    response = requests.get(url)
    with open(f"images\\image_{index}.png", "wb") as f:
        f.write(response.content)

#### 3. Функция последовательного скачивания.

In [3]:
def download_images(urls): # скачиваем изображение из списка URL
    for index, url in enumerate(urls):
        download_image(url, index)

#### 4. Функция скачивания с помощью threading (параллелизм потоков).

In [4]:
def threaded_images(urls): # скачиваем изображение из списка URL с использованием потоков и измеряем время выполнения.
    threads = []
    for index, url in enumerate(urls):
        thread = threading.Thread(target=download_image, args=(url, index))
        threads.append(thread)
        thread.start() # запускаем поток
    for t in threads: # ожидаем завершения всех потоков
        t.join()
    threads = []

#### 5. Функция скачивания с помощью multiprocessing (параллелизм процессов).

In [5]:
def multiprocessing_images(urls):
    with concurrent.futures.ProcessPoolExecutor() as executor:
        executor.map(download_image, [(url, index) for index, url in enumerate(urls)])

#### 6. Реализация скачивания.

In [6]:
if __name__ == "__main__":
    start_time_1 = time.time()  # запоминаем время начала
    download_images(images)
    end_time_1 = time.time()  # запоминаем время окончания
    download_time = end_time_1 - start_time_1  # вычисляем время выполнения
    print(f"Общее время последовательного скачивания: {download_time:.2f} секунд") # выводим 2 знака после запятой тип float

    start_time_2 = time.time() 
    threaded_images(images)
    end_time_2 = time.time()  
    threaded_time = end_time_2 - start_time_2  
    print(f"Общее время скачивания с помощью потоков: {threaded_time:.2f} секунд")

    start_time_3 = time.time()  
    multiprocessing_images(images)
    end_time_3 = time.time()  
    processing_time = end_time_3 - start_time_3  
    print(f"Общее время скачивания с помощью процессов: {processing_time:.2f} секунд")

Общее время последовательного скачивания: 78.49 секунд
Общее время скачивания с помощью потоков: 4.94 секунд
Общее время скачивания с помощью процессов: 0.19 секунд


#### 7. Выводы.

Реализовали скачивание картинок из интернета с помощью threading, multiprocessing и последовательного скачивания. 
Сравнили время скачивания картинок.
Скачивая 100 картинок по заданным ссылкам, получили  
- время последовательного скачивания  
1 тест: 65.88 секунд,  
2 тест: 63.57 секунд,  
3 тест: 78.49 секунд, 
- время скачивания с помощью потоков: 11.40 секунд, 5.04 секунд, 4.94 секунды
- время скачивания с помощью процессов: 0.22 секунды, 0.20 секунды, 0.19 секунды